python_modules/dagster/dagster/_core/definitions/multi_asset_sensor_definition.py

```diff
@@ -440,26 +441,40 @@ def materialization_records_for_key(
             asset_key (AssetKey): The asset to fetch materialization events for
             limit (Optional[int]): The number of events to fetch
         """
-        from dagster._core.events import DagsterEventType
-        from dagster._core.storage.event_log.base import EventRecordsFilter
+        from dagster._core.event_api import AssetRecordsFilter
 
         asset_key = check.inst_param(asset_key, "asset_key", AssetKey)
         if asset_key not in self._assets_by_key:
             raise DagsterInvalidInvocationError(f"Asset key {asset_key} not monitored by sensor.")
 
-        events = list(
-            self.instance.get_event_records(
-                EventRecordsFilter(
-                    event_type=DagsterEventType.ASSET_MATERIALIZATION,
-                    asset_key=asset_key,
-                    after_cursor=self._get_cursor(asset_key).latest_consumed_event_id,
-                ),
-                ascending=True,
-                limit=limit,
-            )
-        )
+        if not limit:
+            deprecation_warning("Calling materialization_records_for_key without a limit", "1.8")
+            records = []
+            has_more = True
+            cursor = None
+            while has_more:
+                result = self.instance.fetch_materializations(
+                    AssetRecordsFilter(
+                        asset_key=asset_key,
+                        after_storage_id=self._get_cursor(asset_key).latest_consumed_event_id,
+                    ),
+                    ascending=True,
+                    limit=FETCH_MATERIALIZATION_BATCH_SIZE,
+                    cursor=cursor,
+                )
+                cursor = result.cursor
+                has_more = result.has_more
+                records.extend(result.records)
+            return records
 
-        return events
+        return self.instance.fetch_materializations(
```

Comment:
not blocking, but there's probably a way to avoid duplicate code here, e.g. by tracking the total number of records we've fetched so far across all iterations and seeing when it reaches the user-provided limit

Reply:
yeah... this is the PR that I think could be rewritten the most.  I haven't quite grokked all the fetching patterns - it feels like there's a better way to do this.

Also, it feels like we should revisit some of the APIs that we expose here to have better data fetching hygiene.  Probably not the highest priority item, but if you had time to pair on this, I'd definitely be willing.

Reply:
I wouldn't be surprised if we remove multi-asset sensor after auto-materialize policies support user-defined functions